### ISSUE TITLE
[PATCH product-offers] Błąd w niektórych produktach, w innych wszystko jest w porządku

### ISSUE THREAD

**Issue by vitalijr2** [allegro/allegro-api#8636]
Zmieniam product-offer dodając produkt, robię to poprzez PATCH do zasobu https://api.allegro.pl/sale/product-offers/******

Wszystko działa dla kilkuset produktów, nie można zaktualizować jednego z powodu błędu „Podana wartość parametru 'Marka'(25929) (inna(25929_41)) nie odpowiada istniejącej wartości parametru (Peterson(25929_843987))"

Wraz z PATCH przesyłam pełną kopię istniejącego produktu i nowy obok niego
Może robię to źle i muszę wysłać tylko identyfikatory productów?

**Comment by vitalijr2** [comment]
Minimalny patch jest:
```json
{
   "productSet":[
      {
         "product":{
            "id":"abc-123-qwerty"
         }
      },
      {
         "product":{
            "id":"xyz-789-zxcvbn"
         }
      }
   ]
}
```

**Comment by PrzemyslawLukanowski** [comment]
Podaj proszę numer oferty lub trace-id wywołania, dla którego otrzymujesz ten błąd. Przyjrzymy się temu.

**Comment by vitalijr2** [comment]
Dziękuje,

nr. oferty `15290537015`

Wypróbowałem minimalny patch, w którem produkty mają tylko identyfikator  - działa poprawnie

Więc w zasadzie problem można uznać za rozwiązany.

**Comment by PrzemyslawLukanowski** [comment]
Ok, dziękuję za informację. Dodam tylko, że problemem było to, że w parametrze marka przekazywałeś inną wartość (25929_41) niż ta, która widnieje w produkcie (25929_843987) połączonym z ofertą.

**Comment by vitalijr2** [comment]
**Update:**
[tak jak pisałem](https://github.com/allegro/allegro-api/issues/8636#issuecomment-1996595687), teraz w patchu używam tylko identyfikatorów produktów

ale czasami nadal pojawiają się błędy, że produkt nie posiada niezbędnych parametrów

- trace-id: 231904a58b0e7f70
  userMessage: Uzupełnij parametry obowiązkowe: Typ, Marka, Kolor.
  message: Missing required parameters: 218093, 127788, 4906
- trace-id: 56b08eea0bd54759, 5cbc56141a8dd248, c2fea3300ce788ec
  userMessage: Musisz dodać przynajmniej 1 zdjęcie.
  message: You have to add at least 1 image.
- trace-id: 6d0f9ae8e18b7bd8
  userMessage: The user does not have Returns Terms named: default.
  message: Unprocessable Entity
- trace-id: c1bc40714f2692bd
  userMessage: Uzupełnij parametry obowiązkowe: EAN (GTIN).
  message: Missing required parameters: 225693
- trace-id: 2dfc0394ded5562
  userMessage: Uzupełnij parametry obowiązkowe: Pojemność, Rodzaj, Marka.
  message: Missing required parameters: 129319, 129318, 211282
- trace-id: fc7379fe81cc163e, 3a46fac74d6b9237
  userMessage: Uzupełnij parametry obowiązkowe: Liczba elementów w zestawie, Linia.
  message: Missing required parameters: 24089, 237214
- trace-id: 76b0a0da5df09dd
  userMessage: Uzupełnij parametry obowiązkowe: EAN (GTIN), Kolor dominujący, Rozmiar (cale), Maksymalna waga użytkownika.
  message: Missing required parameters: 225693, 205826, 247313, 130216
- trace-id: 4525fbbe637b10af
  userMessage: Uzupełnij parametry obowiązkowe: Wiek dziecka, Marka, Seria, Materiał.
  message: Missing required parameters: 3475, 127812, 237222, 18136

**Comment by vitalijr2** [comment]
Jedna z ofert produktów w swoim zestawie produktów posiada produkt o identyfikatorze `null`

Czy możesz mi powiedzieć, co może być tego przyczyną?

**Comment by PrzemyslawLukanowski** [comment]
> ale czasami nadal pojawiają się błędy, że produkt nie posiada niezbędnych parametrów

Tak, z tego co widzę na pierwszym przykładzie to produkty, które masz przypisane do oferty, pochodzą z innej kategorii niż kategoria oferty. W kategorii oferty wymagamy parametrów 218093, 127788, 4906, dlatego musisz je dodatkowo uzupełnić na poziomie `productSet.product[0].parameters`.

> trace-id: 56b08eea0bd54759, 5cbc56141a8dd248, c2fea3300ce788ec
> userMessage: Musisz dodać przynajmniej 1 zdjęcie.
> message: You have to add at least 1 image.

W tym przypadku produkt nie posiada żadnego zdjęcia, dlatego również musisz je dodatkowo uzupełnić po swojej stronie.

> trace-id: 6d0f9ae8e18b7bd8
> userMessage: The user does not have Returns Terms named: default.
> message: Unprocessable Entity

Jak w treści - wygląda na to, że użytkownik nie ma zdefiniowanych warunków zwrotów o nazwie "default". Wspominamy o tym w poradniku:

<img width="514" alt="Zrzut ekranu 2024-03-15 o 08 34 35" src="https://github.com/allegro/allegro-api/assets/51473661/fb703e4a-c842-442a-94c4-25478ad2c116">

> wszystkie tego typu produkty mają publication.status = NOT_LISTED

Dokładnie, oznacza to, że oferta nie jest połączona w tym miejscu z żadnym produktem z katalogu.

**Comment by vitalijr2** [comment]
kolejne pytanie: czy da się tu jakoś zrobić analogię **staging**, czyli tak, żeby zmieniony produkt nie był od razu rejestrowany i pokazywany, ale wpadał do osobnej grupy lub kolejki, która może być potem przez osobę sprawdzona i dopiero po zatwierdzenie zmian, czy będą one wyświetlane w asortymencie?

**Comment by Lukasz-Zurek** [comment]
> Zauważyliśmy kolejny problem ze zmianą opisu głównego produktu w ofercie. Jeśli przy dodawaniu produktu do oferty posłużę się wyłącznie identyfikatorami w patch-u, czasami zmienia się opis:
> 
> * zastępuje obrazy innymi, czasem te zdjęcia w ogóle nie odpowiadają produktowi (np. zestaw zabawek ze zwierzętami otrzymał obraz zabawkowego akordeonu),
> * zastępuje opis produktu
> 
> Mam następujące wymagania:
> 
> * zmiana zdjęć w ofertach (muszą być oryginalne)
> * zmiana opisów (muszą być oryginalne)
> * produkt musi być połączony z katalogiem Allegro, natomiast NIE może pobierać opisów i zdjęć z katalogu - muszą pozostać oryginalne, tak jak były pierwotnie
> 
> 
> 
> **W interfejsie zarządzania asortymentem dostępna jest funkcja „dodaj zdjęcie i opis z katalogu” – przy modyfikowaniu naszych ofert przez API chcielibyśmy uniknąć tej funkcji**

Zapoznaj się proszę z tym [poradnikiem](https://developer.allegro.pl/tutorials/jak-jednym-requestem-wystawic-oferte-powiazana-z-produktem-D7Kj9gw4xFA#wlasne-zdjecia-i-opis-oferty). Czyli jeśli przykładowo chcesz przekazać tylko własne (oryginalne) zdjęcia (w sekcji "images", poza "productSet") i opis w "description", przekazujesz **"product.id"**, a w **"product.images"** przekazujesz pustą tablicę. Jeśli zatem chcesz, aby w ofercie były wyłącznie twoje zdjęcia, dodaj swój własny opis.

> Jeżeli w patch-e nie będę posługiwał się identyfikatorem, a pełnym opisem produktu, czy mogę zagwarantować spełnienie warunków?

Jeżeli na powyższym zasobie nie będziesz posługiwał się identyfikatorem produktu, przekaż dane, które pozwolą nam rozpoznać produkt.

> Czy można w jakiś sposób sprawdzić, co zostało dodane do oferty automatycznie?

Nie, aczkolwiek w odpowiedzi na edycję, zawsze zwracamy wszystkie te dane, które są dostępne w ofercie (po edycji).
>
>kolejne pytanie: czy da się tu jakoś zrobić analogię staging, czyli tak, żeby zmieniony produkt nie był od razu rejestrowany i pokazywany, ale wpadał do osobnej grupy lub kolejki, która może być potem przez osobę sprawdzona i dopiero po zatwierdzenie zmian, czy będą one wyświetlane w asortymencie?

Zasób [PATCH /sale/product-offers/{offerId}](https://developer.allegro.pl/documentation/#operation/editProductOffers) służy do edycji oferty, a nie produktu, więc tu takiej możliwości nie ma. Natomiast jest jeszcze zasób do zgłaszania błędów w produkcie [POST /sale/products/{productId}/change-proposals](https://developer.allegro.pl/documentation/#operation/productChangeProposal), gdzie dla wskazanego produktu przekazujesz np. parametry, które według Ciebie są błędne. Takie żądanie nie jest od razu przetwarzane, ponieważ trafia ono do zespołu odpowiedzialnego za katalog produktów, który to weryfikuje.
